File: src/1425.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int constrainedSubsetSum(int* nums, int numsSize, int k) {
    int retVal = 0;

    struct deque {
        int value;
        int index;
    } monotonicDeque[numsSize];
    int monotonicDequeHead = 0;
    int monotonicDequeTail = 0;
    monotonicDeque[monotonicDequeTail].value = nums[0];
    monotonicDeque[monotonicDequeTail].index = 0;
    monotonicDequeTail++;

    int dp[numsSize];
    memset(dp, 0, sizeof(dp));
    dp[0] = nums[0];

    retVal += nums[0];

    int i;
    for (i = 1; i < numsSize; i++) {
        dp[i] = fmax(0, dp[monotonicDeque[monotonicDequeHead].index]) + nums[i];
        retVal = fmax(retVal, dp[i]);

        while ((monotonicDequeHead < monotonicDequeTail) &&
               (dp[i] >= dp[monotonicDeque[monotonicDequeTail - 1].index])) {
            monotonicDequeTail--;
        }
        monotonicDeque[monotonicDequeTail].value = nums[i];
        monotonicDeque[monotonicDequeTail].index = i;
        monotonicDequeTail++;
        if ((i - monotonicDeque[monotonicDequeHead].index) >= k) {
            monotonicDequeHead++;
        }
    }

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (500)
    struct testCaseType {
        int nums[MAX_SIZE];
        int numsSize;
        int k;
    } testCase[] = {{{10, 2, -10, 5, 20}, 5, 2}, {{-1, -2, -3}, 3, 1}, {{10, -2, -10, -5, 20}, 5, 2}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: nums = [10,2,-10,5,20], k = 2
     *  Output: 37
     *
     *  Input: nums = [-1,-2,-3], k = 1
     *  Output: -1
     *
     *  Input: nums = [10,-2,-10,-5,20], k = 2
     *  Output: 23
     */

    int answer;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: nums = [");
        for (j = 0; j < testCase[i].numsSize; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", testCase[i].nums[j]);
        }
        printf("], k = %d\n", testCase[i].k);

        answer = constrainedSubsetSum(testCase[i].nums, testCase[i].numsSize, testCase[i].k);
        printf("Output: %d\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}

File: src/1425.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def constrainedSubsetSum(self, nums: List[int], k: int) -> int:
        retVal = 0

        numsSize = len(nums)

        monotonicDeque = deque()
        dp = [0] * numsSize

        for i in range(numsSize):
            if monotonicDeque and i - monotonicDeque[0] > k:
                monotonicDeque.popleft()

            dp[i] = (dp[monotonicDeque[0]] if monotonicDeque else 0) + nums[i]
            while monotonicDeque and dp[monotonicDeque[-1]] < dp[i]:
                monotonicDeque.pop()

            if dp[i] > 0:
                monotonicDeque.append(i)

        retVal = max(dp)

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums, k in zip([[10, 2, -10, 5, 20], [-1, -2, -3], [10, -2, -10, -5, 20]], [2, 1, 2]):
            # /* Example
            #  *  Input: nums = [10,2,-10,5,20], k = 2
            #  *  Output: 37
            #  *
            #  *  Input: nums = [-1,-2,-3], k = 1
            #  *  Output: -1
            #  *
            #  *  Input: nums = [10,-2,-10,-5,20], k = 2
            #  *  Output: 23
            #  */
            logging.info("Input: nums = %s, k = %s", nums, k)

            retVal = pSolution.constrainedSubsetSum(nums, k)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
